### PR TITLE
StackExchangeRedis instrumentation: Dependency injection support

### LINKED ISSUE #2011: Instrumentation.StackExchangeRedis Should Resolve IConnectionMultiplexer from DI
# Feature Request

I have registered `IConnectionMultiplexer` in my DI container for health checks, a Redis implementation of `IDistributedCache` and other reasons. It would be nice if `OpenTelemetry.Instrumentation.StackExchangeRedis` contained an overload for the `AddRedisInstrumentation` function that allows you to resolve `IConnectionMultiplexer` from DI:

```cs
public static TracerProviderBuilder AddRedisInstrumentation(
    this TracerProviderBuilder builder,
    Func<IConnectionMultiplexer, IServiceBuilder> getConnection,
    Action<StackExchangeRedisCallsInstrumentationOptions> configureOptions = null);
```

Another overload that assumes that `IConnectionMultiplexer` is registered in DI would also be helpful:

```cs
public static TracerProviderBuilder AddRedisInstrumentation(
    this TracerProviderBuilder builder,
    Action<StackExchangeRedisCallsInstrumentationOptions> configureOptions = null);
```

**Is your feature request related to a problem?**

I have to create a new connection specifically for Open Telemetry.

**Describe the solution you'd like:**

I would like to reuse the connection I'm using elsewhere.

### PR DESCRIPTION
**Pull request by CodeBlanch**
Fixes #2011

## Changes

Updated StackExchangeRedis instrumentation `TracerProviderBuilder` extension to resolve `IConnectionMultiplexer` & `StackExchangeRedisCallsInstrumentationOptions` through DI if it is available.

## TODOs

* [X] `CHANGELOG.md` updated for non-trivial changes
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by reyang — APPROVED** [review]
LGTM.

File: test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/StackExchangeRedisCallsInstrumentationTests.cs [hunk 1]
```diff
@@ -203,13 +204,44 @@ public void StackExchangeRedis_BadArgs()
...
+            connectionOptions.EndPoints.Add("localhost");
+
+            var services = new ServiceCollection();
+            services.AddSingleton<IConnectionMultiplexer>(ConnectionMultiplexer.Connect(connectionOptions));
+
+            services.AddOpenTelemetryTracing(builder => builder.AddRedisInstrumentation());
+
+            using var serviceProvider = services.BuildServiceProvider();
+
+            var tracerProvider = serviceProvider.GetRequiredService<TracerProvider>();
```
Comment (utpilla) [inline test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/StackExchangeRedisCallsInstrumentationTests.cs:230]: Do we want to test that the `DeferredTracerProviderBuilder` is correctly adding `RedisInstrumentation` by assigning some variable in `configure` action and assert that fact. Something like this: https://github.com/open-telemetry/opentelemetry-dotnet/blob/2347c48203492d05716d580f8310c94057d3195c/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/DependencyInjectionConfigTests.cs#L70

Comment (CodeBlanch) [reply test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/StackExchangeRedisCallsInstrumentationTests.cs:240]: Updated

**Review by utpilla — APPROVED** [review]

**Review by cijothomas — APPROVED** [review]
LGTM

**Pull request merged**